starlingcaptureapi/fs_watcher.py

```diff
@@ -27,106 +29,163 @@ def caught_and_logged_exceptions(event):
 class FsWatcher:
     """Watches directories for file changes."""
 
-    def __init__(self, organization_id):
-        self.organization_id = organization_id
+    def __init__(self, org_config: dict, collection_id: str = None):
+        self.org_config = org_config
+        self.collection_id = collection_id
+        self.organization_id = org_config.get("id")
+        self.asset_helper = AssetHelper(self.organization_id)
+        self.observer = Observer()
+
+    @staticmethod
+    def start(org_config: dict):
+        FsWatcher(org_config).watch()
+
+    @staticmethod
+    def init_all(
+        all_org_config: config.OrganizationConfig = config.ORGANIZATION_CONFIG,
+    ) -> list[multiprocessing.Process]:
+        """Initialize file watcher processes for the given configuration.
 
-    def watch(self):
-        """
         Args:
-            organization_id: a string identifying the organization we are watching for
+            org_config: configuration for all organizations and their actions
+
+        Returns:
+            list of un-started processes containing FsWatcher instances
         """
-        observer = Observer()
-        patterns = ["*.jpg", "*.jpeg"]
-        asset_helper = AssetHelper(self.organization_id)
-        observer.schedule(
-            self.AddHandler(patterns=patterns).set_org_id(self.organization_id),
-            recursive=True,
-            path=asset_helper.get_assets_add(),
+        procs = []
+        for org_id, org_config in all_org_config.config.items():
+            procs.append(
+                multiprocessing.Process(
+                    name=f"fs_watcher_{org_id}",
+                    target=FsWatcher.start,
+                    args=(org_config,),
+                )
+            )
+        return procs
+
+    def watch(self):
+        """Start file watching handlers."""
+        self._schedule_legacy_handlers()
+        for collection_id, collection_config in self.org_config.get(
+            "collections", {}
+        ).items():
+            patterns = [
+                f"*.{ext}"
+                for ext in collection_config.get("conf", {}).get("asset_extensions", [])
+            ]
```

Comment:
I am assuming that the configuration for an archive action will set the file extensions to watch to `.zip`. Seems better to me to rely on each action configuration to provide such information, instead of special-casing some actions to have hardcoded extensions to watch for.

The same goes for directories to watch: it would be good to be consistent about how we do this across all actions. We could have some logic that uses a "default path" for each action (based on the action name), but if the configuration provides a path to watch, then we could watch that. Not sure if the added complexity is worth it, just to have slightly different input/output directories to watch. Please confirm @benhylau which way would you like to do with this.

Reply:
>I am assuming that the configuration for an archive action will set the file extensions to watch to .zip.

The config.json defines the _asset_ ext, not the ext of the package that gets dropped off atomically for processing. So this is always `.zip`. When it gets unzipped, you'll find:

3. The ZIP contains three files, named as such:
  - `sha256(asset).ext`: the asset file with `ext` matching one of the Collection's `asset_extensions`
  - `sha256(asset)-meta-content.json`: the metadata associated with the asset file
  - `sha256(asset)-meta-recorder.json`: the metadata associated with the recorder of the asset

The config's `asset_extensions` array refers to that. So the watch should always be on `.zip`.

>Seems better to me to rely on each action configuration to provide such information, instead of special-casing some actions to have hardcoded extensions to watch for.

Not special-casing. We'll have to change the capture app's JPG delivery to zip up the JPG alongside two `.json` meta files too.

>The same goes for directories to watch: it would be good to be consistent about how we do this across all actions.

I think we're standardizing on: `./{collection_id}/input/*.zip`

Reply:
@benhylau does this mean that all the handlers for all the actions in this PR should be watching for `*.zip`? (If so, this will break the existing action implementations.)

Or are we special-casing the `archive` action in this PR, and watching for `*.zip` for that action only?

As for the input directories to watch, are you saying that all handlers for all actions for a given collection should watch the same input directory, at `./{collection_id}/input/*.zip`? Or did you mean each action will have its own input sub-directory, like `./{collection_id}/action-name/input/*.zip`?

Reply:
@benhylau I addressed this as we discussed in Slack: the `archive` action watches `collection_id/input/*.zip`, and all the other actions keep the legacy behavior. I added comments in the code talking about how the legacy behavior should be changed, and linking to the issue that tracks that (https://github.com/starlinglab/starling-integrity-api/issues/79).